crates/sui-json-rpc/src/authority_state.rs

```diff
@@ -0,0 +1,654 @@
+        transaction_digest: &TransactionDigest,
+    ) -> StateReadResult<Option<Vec<(ObjectID, SequenceNumber)>>> {
+        Ok(self.loaded_child_object_versions(transaction_digest)?)
+    }
+
+    fn get_chain_identifier(&self) -> StateReadResult<ChainIdentifier> {
+        Ok(self
+            .get_chain_identifier()
+            .ok_or(anyhow!("Chain identifier not found"))?)
+    }
+}
+
+/// This implementation allows `S` to be a dynamically sized type (DST) that implements ObjectProvider
+/// Valid as `S` is referenced only, and memory management is handled by `Arc`
+#[async_trait]
+impl<S: ?Sized + State> ObjectProvider for Arc<S> {
+    type Error = StateReadError;
+
+    async fn get_object(
+        &self,
+        id: &ObjectID,
+        version: &SequenceNumber,
+    ) -> Result<Object, Self::Error> {
+        Ok(self.get_past_object_read(id, *version)?.into_object()?)
+    }
+
+    async fn find_object_lt_or_eq_version(
+        &self,
+        id: &ObjectID,
+        version: &SequenceNumber,
+    ) -> Result<Option<Object>, Self::Error> {
+        let database = self.get_db();
+        let id = *id;
+        let version = *version;
+        spawn_monitored_task!(async move { database.find_object_lt_or_eq_version(id, version) })
+            .await
+            .map_err(StateReadError::from)
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum StateReadInternalError {
+    #[error(transparent)]
+    SuiError(#[from] SuiError),
+    #[error(transparent)]
+    JoinError(#[from] JoinError),
+    #[error(transparent)]
+    Anyhow(#[from] anyhow::Error),
+}
+
+#[derive(Debug, Error)]
+pub enum StateReadClientError {
+    #[error(transparent)]
+    SuiError(#[from] SuiError),
+    #[error(transparent)]
+    UserInputError(#[from] UserInputError),
+}
+
+/// `StateReadError` is the error type for callers to work with.
+/// It captures all possible errors that can occur while reading state, classifying them into two categories.
+/// Unless `StateReadError` is the final error state before returning to caller, the app may still want error context.
+/// This context is preserved in `Internal` and `Client` variants.
+/// Also implements `From` trait to convert into `RpcError` if this is indeed the final error state.
+#[derive(Debug, Error)]
+pub enum StateReadError {
+    // sui_json_rpc::Error will do the final conversion to generic error message
+    #[error(transparent)]
+    Internal(#[from] StateReadInternalError),
+
+    // Client errors
+    #[error(transparent)]
+    Client(#[from] StateReadClientError),
+}
+
+impl From<SuiError> for StateReadError {
+    fn from(e: SuiError) -> Self {
+        match e {
+            SuiError::IndexStoreNotAvailable
+            | SuiError::TransactionNotFound { .. }
+            | SuiError::UnsupportedFeatureError { .. }
+            | SuiError::UserInputError { .. }
+            | SuiError::WrongMessageVersion { .. } => StateReadError::Client(e.into()),
+            _ => StateReadError::Internal(e.into()),
```

Comment:
It may be useful to clarify somehow when the error is transient, indicating to the user that they can retry, vs an error that is non-transient

Reply:
Oh yeah good call!